File: Rock.JavaScript.Obsidian.Blocks/src/Example/gridTest.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//

import { useConfigurationValues, useInvokeBlockAction } from "@Obsidian/Utility/block";
import { computed, defineComponent, PropType, VNode } from "vue";
import { RockDateTime } from "@Obsidian/Utility/rockDateTime";
import Grid from "./GridTest/grid.partial";
import { AttributeColumnDefinition, GridColumnDefinition, GridData, GridDefinition } from "./GridTest/types";

const Column = defineComponent({
    props: {
        name: {
            type: String as PropType<string>,
            default: ""
        },

        title: {
            type: String as PropType<string>,
            required: false
        },

        sortField: {
            type: String as PropType<string>,
            required: false
        },

        sortValue: {
            type: Object as PropType<((row: Record<string, unknown>, column: GridColumnDefinition) => string | number | undefined)>,
            required: false
        }
    }
});

const dateColumnValueComponent = defineComponent({
    props: {
        column: {
            type: Object as PropType<GridColumnDefinition>,
            required: true
        },
        row: {
            type: Object as PropType<Record<string, unknown>>,
            required: true
        }
    },

    setup(props) {
        const formattedValue = computed(() => {
            if (props.column.field) {
                return RockDateTime.parseISO(props.row[props.column.field] as string)?.toASPString("d") ?? "";
            }
            else {
                return "";
            }
        });

        return {
            formattedValue
        };
    },

    template: `{{ formattedValue }}`
});

const DateColumn = defineComponent({
    props: {
        name: {
            type: String as PropType<string>,
            default: ""
        },

        title: {
            type: String as PropType<string>,
            required: false
        },

        sortField: {
            type: String as PropType<string>,
            required: false
        },

        sortValue: {
            type: Object as PropType<((row: Record<string, unknown>, column: GridColumnDefinition) => string | number | undefined)>,
            required: false
        },

        format: {
            type: Object as PropType<VNode>,
            required: false,
            default: dateColumnValueComponent
        }
    }
});

const booleanColumnValueComponent = defineComponent({
    props: {
        column: {
            type: Object as PropType<GridColumnDefinition>,
            required: true
        },
        row: {
            type: Object as PropType<Record<string, unknown>>,
            required: true
        }
    },

    setup(props) {
        const isTrue = computed(() => {
            if (props.column.field) {
                return props.row[props.column.field] === true;
            }
            else {
                return false;
            }
        });

        return {
            isTrue
        };
    },

    template: `<i v-if="isTrue" class="fa fa-check"></i>`
});

function booleanColumnSortValue(row: Record<string, unknown>, column: GridColumnDefinition): number | undefined {
    if (!column.sortField) {
        return undefined;
    }

    return row[column.sortField] === true ? 1 : 0;
}


const BooleanColumn = defineComponent({
    props: {
        name: {
            type: String as PropType<string>,
            default: ""
        },

        title: {
            type: String as PropType<string>,
            required: false
        },

        sortField: {
            type: String as PropType<string>,
            required: false
        },

        sortValue: {
            type: Object as PropType<((row: Record<string, unknown>, column: GridColumnDefinition) => string | number | undefined)>,
            required: false,
            default: booleanColumnSortValue
        },

        format: {
            type: Object as PropType<VNode>,
            required: false,
            default: booleanColumnValueComponent
        }
    }
});

const AttributeColumns = defineComponent({
    components: {
        Column
    },

    props: {
        attributes: {
            type: Array as PropType<AttributeColumnDefinition[]>,
            default: []
        },

        __attributeColumns: {
            type: Boolean as PropType<boolean>,
            default: true
        }
    }
});

export default defineComponent({
    name: "Example.GridTest",

    components: {
        AttributeColumns,
        BooleanColumn,
        Column,
        DateColumn,
        Grid
    },

    setup() {
        const configuration = useConfigurationValues<{ gridDefinition: GridDefinition }>();
        const invokeBlockAction = useInvokeBlockAction();

        const attributeColumns = computed((): AttributeColumnDefinition[] => {
            return configuration.gridDefinition.attributeColumns ?? [];
        });

        const loadGridData = async (): Promise<GridData> => {
            const result = await invokeBlockAction<GridData>("GetGridData");

            if (result.isSuccess && result.data) {
                return {
                    rows: result.data.rows
                };
            }
            else {
                throw new Error(result.errorMessage ?? "Unknown error while trying to load grid data.");
            }
        };

        function formatDate(value?: string): string {
            if (!value) {
                return "";
            }

            const dt = RockDateTime.parseISO(value);

            return dt?.toASPString("g") ?? "";
        }

        return {
            attributeColumns,
            formatDate,
            gridData: loadGridData
        };
    },

    template: `
<Grid :data="gridData">
    <Column name="name"
            title="Name"
            sortValue="{{ row.name.lastName }} {{ row.name.firstName }}">
        <template #body="{ row }">{{ row.name.firstName }} {{ row.name.lastName }}</template>
    </Column>

    <Column name="email"
            title="Email"
            field="email"
            sortField="email" />

    <Column name="enteredDateTime"
            title="Entered On"
            field="enteredDateTime"
            sortField="enteredDateTime">
        <template #body="{ row }">{{ formatDate(row.enteredDateTime) }}</template>
    </Column>

    <DateColumn name="expirationDateTime"
                title="Expires"
                field="expirationDateTime"
                sortField="expirationDateTime" />

    <BooleanColumn name="isUrgent"
                   title="Urgent"
                   field="isUrgent"
                   sortField="isUrgent" />

    <BooleanColumn name="isPublic"
                   title="Public"
                   field="isPublic"
                   sortField="isPublic" />

    <AttributeColumns :attributes="attributeColumns" />
</Grid>
`
});

File: Rock.Blocks/Example/GridTest.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data.Entity;
using System.Linq;
using System.Linq.Expressions;

using Rock.Attribute;
using Rock.Blocks.Types.Mobile.Prayer;
using Rock.Data;
using Rock.Model;
using Rock.Web.Cache;

namespace Rock.Blocks.Example
{
    /// <summary>
    /// Allows testing the new Grid component.
    /// </summary>
    /// <seealso cref="Rock.Blocks.RockObsidianBlockType" />

    [DisplayName( "Grid Test" )]
    [Category( "Example" )]
    [Description( "Allows testing the new Grid component." )]
    [IconCssClass( "fa fa-list" )]

    [Rock.SystemGuid.EntityTypeGuid( "1934a378-57d6-44d0-b7cd-4443f347a1ee" )]
    public class GridTest : RockObsidianBlockType
    {
        public override string BlockFileUrl => $"{base.BlockFileUrl}.vue";

        public override object GetObsidianBlockInitialization()
        {
            return new
            {
                GridDefinition = GetGridBuilder( GetGridAttributes() ).BuildDefinition()
            };
        }

        [BlockAction]
        public BlockActionResult GetGridData()
        {
            using ( var rockContext = new RockContext() )
            {
                var count = RequestContext.GetPageParameter( "count" )?.AsIntegerOrNull() ?? 10_000;

                var gridAttributes = GetGridAttributes();
                var gridAttributeIds = gridAttributes.Select( a => a.Id ).ToList();

                var sw = System.Diagnostics.Stopwatch.StartNew();
                var prayerRequests = new PrayerRequestService( rockContext )
                    .Queryable()
                    .AsNoTracking()
                    .Take( count )
                    .ToList();
                sw.Stop();
                System.Diagnostics.Debug.WriteLine( $"Entity load took {sw.Elapsed.TotalMilliseconds}ms." );

                sw.Restart();
                Helper.LoadFilteredAttributes( prayerRequests, rockContext, a => gridAttributeIds.Contains( a.Id ) );
                sw.Stop();
                System.Diagnostics.Debug.WriteLine( $"Attribute load took {sw.Elapsed.TotalMilliseconds}ms." );

                sw.Restart();
                var rows = prayerRequests
                    .Select( pr =>
                    {
                        var row = new Dictionary<string, object>
                        {
                            ["name"] = new { pr.FirstName, pr.LastName },
                            ["email"] = pr.Email,
                            ["enteredDateTime"] = pr.EnteredDateTime,
                            ["expirationDateTime"] = pr.ExpirationDate,
                            ["isUrgent"] = pr.IsUrgent,
                            ["isPublic"] = pr.IsPublic
                        };

                        foreach ( var attrValue in pr.AttributeValues )
                        {
                            row[$"attr_{attrValue.Key}"] = pr.GetAttributeCondensedHtmlValue( attrValue.Key );
                        }

                        return row;
                    } )
                    .ToList();

                sw.Stop();
                System.Diagnostics.Debug.WriteLine( $"Manual row translation took {sw.Elapsed.TotalMilliseconds}ms." );

                sw.Restart();
                rows = GetGridBuilder( gridAttributes ).BuildRows( prayerRequests );

                sw.Stop();
                System.Diagnostics.Debug.WriteLine( $"Row translation took {sw.Elapsed.TotalMilliseconds}ms." );

                var definition = GetGridBuilder( gridAttributes ).BuildDefinition();

                var gridData = new
                {
                    Rows = rows
                };

                return ActionOk( gridData );
            }
        }

        private List<AttributeCache> GetGridAttributes()
        {
            var entityTypeId = EntityTypeCache.GetId<PrayerRequest>().Value;

            return AttributeCache.GetByEntityTypeQualifier( entityTypeId, string.Empty, string.Empty, false )
                .Where( a => a.IsGridColumn )
                .OrderBy( a => a.Order )
                .ThenBy( a => a.Name )
                .ToList();
        }

        private GridBuilder<PrayerRequest> GetGridBuilder( List<AttributeCache> gridAttributes )
        {
            return new GridBuilder<PrayerRequest>()
                .AddField( "name", pr => new { pr.FirstName, pr.LastName } )
                .AddTextField( "email", pr => pr.Email )
                .AddDateTimeField( "enteredDateTime", pr => pr.EnteredDateTime )
                .AddDateTimeField( "expirationDateTime", pr => pr.ExpirationDate )
                .AddField( "isUrgent", pr => pr.IsUrgent )
                .AddField( "isPublic", pr => pr.IsPublic )
                .AddAttributeFields( gridAttributes );
        }
    }

    public class GridBuilder<T>
    {
        private readonly Dictionary<string, Func<T, object>> _fieldActions = new Dictionary<string, Func<T, object>>();

        private readonly List<Action<GridDefinition>> _definitionActions = new List<Action<GridDefinition>>();

        public GridBuilder()
        {
        }

        public GridBuilder<T> AddPersonField( string name, Func<T, Person> valueExpression )
        {
            return AddField( name, row =>
            {
                var person = valueExpression( row );

                if ( person == null )
                {
                    return null;
                }

                return new
                {
                    person.FirstName,
                    person.NickName,
                    person.LastName,
                    person.PhotoUrl
                };
            } );
        }

        public GridBuilder<T> AddDateTimeField( string name, Func<T, DateTime?> valueExpression )
        {
            return AddField( name, row => valueExpression( row )?.ToRockDateTimeOffset() );
        }

        public GridBuilder<T> AddTextField( string name, Func<T, string> valueExpression )
        {
            return AddField( name, row => valueExpression( row ) );
        }

        public GridBuilder<T> AddAttributeFields( IEnumerable<AttributeCache> attributes )
        {
            if ( !typeof( IHasAttributes ).IsAssignableFrom( typeof( T ) ) )
            {
                throw new Exception( $"The type '{typeof( T ).FullName}' does not support attributes." );
            }

            foreach ( var attribute in attributes )
            {
                var key = attribute.Key;
                var fieldKey = $"attr_{key}";

                AddField( fieldKey, item =>
                {
                    var attributeRow = item as IHasAttributes;

                    return attributeRow.GetAttributeCondensedHtmlValue( key );
                } );

                _definitionActions.Add( definition =>
                {
                    definition.AttributeColumns.Add( new AttributeColumnDefinition
                    {
                        Name = fieldKey,
                        Title = attribute.Name
                    } );
                } );
            }

            return this;
        }

        public GridBuilder<T> AddField( string name, Func<T, object> valueExpression )
        {
            if ( _fieldActions.ContainsKey( name ) )
            {
            }

            _fieldActions.Add( name, valueExpression );

            return this;
        }

        public List<Dictionary<string, object>> BuildRows( IEnumerable<T> items )
        {
            var fieldKeys = _fieldActions.Keys.ToArray();

            return items
                .Select( item =>
                {
                    var row = new Dictionary<string, object>();

                    for ( int i = 0; i < fieldKeys.Length; i++ )
                    {
                        var key = fieldKeys[i];
                        var value = _fieldActions[key]( item );

                        row[key] = value;
                    }

                    return row;
                } )
                .ToList();
        }

        public GridDefinition BuildDefinition()
        {
            var definition = new GridDefinition
            {
                AttributeColumns = new List<AttributeColumnDefinition>()
            };

            foreach ( var action in _definitionActions )
            {
                action( definition );
            }

            return definition;
        }
    }

    public class GridDefinition
    {
        public List<AttributeColumnDefinition> AttributeColumns { get; set; }
    }

    public class AttributeColumnDefinition
    {
        public string Name { get; set; }

        public string Title { get; set; }
    }

    public class GridData
    {
        public List<Dictionary<string, object>> Rows { get; set; }
    }
}
